src/canopy-sofe-extensions.js

```diff
@@ -1,62 +1,72 @@
-            }
-          })
-          .catch(ex => {
-            throw ex;
-          });
-        } else if (localRegex.test(localStorageValue)) {
-          postLocateNext(`${localDomain}:${localStorageValue}/${serviceName}.js`);
-        } else if (localStorageValue === 'stage') {
-          throw new Error(`'stage' is not a valid override for sofe services. Please change your sofe override for '${serviceName}' to be either 'app-stage' or 'cdn-stage'`);
-        } else {
-          postLocateNext(postLocateLoad);
-        }
+import { applyMiddleware, getManifest, getServiceName, getServiceUrl, InvalidServiceName } from 'sofe';
+
+const envs = {
+  integ: 'https://cdn-integ.canopy.ninja/sofe-manifest.json',
+  'cdn-integ': 'https://cdn-integ.canopy.ninja/sofe-manifest.json',
+  stage: 'https://cdn-stage.canopy.ninja/sofe-manifest.json',
+  'app-stage': 'https://cdn-stage.canopy.ninja/sofe-manifest.json',
+  'cdn-stage': 'https://cdn-stage.canopy.ninja/sofe-manifest.json',
+  prod: 'https://cdn.canopytax.com/sofe-manifest.json',
+};
+
+const localRegex = /^[0-9][0-9][0-9][0-9]$/
+
+const canopyMiddleware = () => (preLocateLoad, preLocateNext) => {
+  const serviceName = getServiceName(preLocateLoad);
+
+  preLocateNext(preLocateLoad);
+
+  return (postLocateLoad, postLocateNext) => {
+    if (serviceName) {
+      const localStorageValue = localStorage.getItem(`sofe:${serviceName}`);
+      if (typeof localStorageValue === 'string' && envs[localStorageValue]) {
+        getManifest(envs[localStorageValue])
+        .then(manifest => {
+          if (manifest[serviceName]) {
+            console.log(`Overriding sofe service '${serviceName}' to use the version found on '${localStorageValue}'`);
+            postLocateNext(manifest[serviceName]);
+          } else {
+            throw new Error(`Cannot find sofe service '${serviceName}' at the sofe manifest for '${localStorageValue}'`);
+          }
+        })
+        .catch(ex => {
+          throw ex;
+        });
+      } else if (localRegex.test(localStorageValue)) {
+        postLocateNext(`https://localhost:${localStorageValue}/${serviceName}.js`);
+      } else if (localStorageValue === 'stage') {
+        throw new Error(`'stage' is not a valid override for sofe services. Please change your sofe override for '${serviceName}' to be either 'app-stage' or 'cdn-stage'`);
       } else {
-        console.warn(`canopy-sofe-extensions not working properly (overrides to "cdn-stage", "thanos", etc) because sofe's getServiceName function returned something falsy for`, preLocateLoad);
         postLocateNext(postLocateLoad);
       }
+    } else {
+      console.warn(`canopy-sofe-extensions not working properly (overrides to "cdn-stage", "thanos", etc) because sofe's getServiceName function returned something falsy for`, preLocateLoad);
+      postLocateNext(postLocateLoad);
     }
   }
+}
+
+applyMiddleware(canopyMiddleware);
+
+export function calculatePublicPath(serviceName) {
```

Comment:
this is the new function

Reply:
The reason we put it into canopy-sofe-extensions is that the webpack public path for our projects doesn't work due to the dynamic nature of our sofe service urls and overrides. So this project made sense, since it's canopy-specific for getting some nice sofe things to work

Reply:
This function will be called inside individual sofe services like this:

```js
__webpack_public_path__ = calculatePublicPath('notifications-ui')
```

which will make code splits work for that service, even before any single-spa application is bootstrapped inside of the service